lib/bucket_store/key_storage.rb

```diff
@@ -60,25 +60,27 @@ def download
     # @param [String] content The content to upload
     # @return [String] The final `key` where the content has been uploaded
     # @example Upload a file
-    #   BucketStore.for("inmemory://bucket/file.xml").upload("hello world")
+    #   BucketStore.for("inmemory://bucket/file.xml").upload!("hello world")
     def upload!(content)
```

Comment:
An alternative approach here is assuming the content is a file then having a `if content.instance_of?(String)` to wrap Strings in StringIO. For the sake of backwards compatibility / explictness I decided against that approach but happy to be convinced that it's a better one

Reply:
That would be backwards compatible, as it doesn't require any changes to existing codebases so the main thing to understand is whether the unified API is a better choice compared to having explicit some separation.

Wonder if the this is a case better served by streaming uploads (see #63): if you use the non-streaming version, then only strings can be uploaded, but file-like objects are handled by `.stream`. This will be the case implicitly here, since under the hood s3/gcs will stream file reads. Though I guess it opens a question of whether the interfaced proposed in #63 is the right one... WDYT? Happy to go either way, though I'd slightly prefer a single `.upload!` method to reduce cognitive load when using the gem

Reply:
So I actually missed that #63 added support for streaming uploads as the PR title only mentioned downloads. I see now there is an upload version with `NotImplemented` on it.

> Though I guess it opens a question of whether the interfaced proposed in https://github.com/gocardless/bucket-store/pull/63 is the right one... WDYT? Happy to go either way, though I'd slightly prefer a single .upload! method to reduce cognitive load when using the gem

I will say that from a "least surprise" perspective having a single `upload!` method that accepts a string or a file like object and handles the difference behind the scenes seems to be a common approach within Ruby e.g. the S3 SDK does this for file uploads vs exposing a separate `put_object_from_file` method / explicit `.stream` version. Naturally there are counter examples to this but want to say the spirit of "do what I mean" seems to prevail within other libraries.

In terms of API consistency I'm also wondering if exposing streaming downoads via an IO like API might fit better with existing libraries e.g. https://github.com/soulcutter/saxerator accepts a file like object for XML parsing.

WDYR? Happy to have a call on this